### PR TITLE
Include Prometheus in default build

### PR DESCRIPTION
**Pull request by andybug**
- Remove `prometheus-metrics` feature flag
- Refactor `prometheus_metrics.rs` to implement the `Collector` trait
- Bumped `actix-web-prom` to the latest version
- Replaced `unwrap()` calls with error handling or at least `expect()`

https://github.com/LemmyNet/lemmy/issues/3558

### CODE REVIEW

File: src/prometheus_metrics.rs [hunk 1]
```diff
@@ -12,24 +10,34 @@ use std::{
...
 }
 
 static DEFAULT_BIND: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
 static DEFAULT_PORT: i32 = 10002;
 
-pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyContext) {
+pub fn serve_prometheus_metrics(config: Option<&PrometheusConfig>, lemmy_context: LemmyContext) {
+  // if there is no prometheus config block, disable the server
+  if config.is_none() {
+    eprintln!("No Prometheus config provided, will not start Prometheus server");
```
Comment (Nutomic) [inline src/prometheus_metrics.rs:28]: Were not using println anywhere, better use `info!`

File: src/lib.rs [hunk 2]
```diff
@@ -164,11 +159,10 @@ pub async fn start_lemmy_server() -> Result<(), LemmyError> {
 
   // this must come before the HttpServer creation
   // creates a middleware that populates http metrics for each path, method, and status code
-  #[cfg(feature = "prometheus-metrics")]
   let prom_api_metrics = PrometheusMetricsBuilder::new("lemmy_api")
     .registry(default_registry().clone())
     .build()
-    .expect("Should always be buildable");
+    .expect("Failed to initialize actix_web_prom metrics");
```
Comment (Nutomic) [inline src/lib.rs:165]: We can move this inside prometheus_metrics.rs to shorten this file a bit.

File: src/prometheus_metrics.rs [hunk 3]
```diff
@@ -60,63 +68,77 @@ pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyC
...
-  // gather metrics from registry and encode in prometheus format
+  // gather metrics from registry
   let metric_families = prometheus::gather();
-  encoder.encode(&metric_families, &mut buffer).unwrap();
-  let output = String::from_utf8(buffer).unwrap();
 
-  HttpResponse::Ok().body(output)
+  // encode in prometheus format
+  if let Err(err) = encoder.encode(&metric_families, &mut buffer) {
+    eprintln!("Prometheus encoding error: {}", err);
```
Comment (Nutomic) [inline src/prometheus_metrics.rs:80]: `warn!` or `err!` for errors

File: src/prometheus_metrics.rs [hunk 4]
```diff
@@ -60,63 +68,77 @@ pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyC
...
+    ))
+    .expect("Prometheus DbPoolMetrics: failed to create max_size metric");
+
+    let available = Gauge::with_opts(Opts::new(
       "lemmy_db_pool_available_connections",
       "Number of available connections in the pool",
     ))
-    .unwrap(),
-  };
+    .expect("Prometheus DbPoolMetrics: failed to create available metric");
```
Comment (Nutomic) [inline src/prometheus_metrics.rs:112]: Doesnt hurt to use `?` operator and return result here (unless that doesnt compile for some reason).

File: src/prometheus_metrics.rs [hunk 5]
```diff
@@ -49,7 +57,7 @@ pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyC
           .route("/metrics", web::get().to(metrics))
       })
       .bind((bind, port as u16))
-      .unwrap_or_else(|_| panic!("Cannot bind to {}:{}", bind, port))
+      .unwrap_or_else(|_| panic!("Prometheus server cannot bind to {}:{}", bind, port))
```
Comment (Nutomic) [inline src/prometheus_metrics.rs:60]: Never swallow errors, include it in the log message.

File: src/prometheus_metrics.rs [hunk 6]
```diff
@@ -60,63 +68,77 @@ pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyC
...
+    return HttpResponse::InternalServerError().finish();
+  }
+
+  // convert to utf-8 and return
+  match String::from_utf8(buffer) {
+    Ok(body) => HttpResponse::Ok().body(body),
+    Err(err) => {
+      eprintln!("Prometheus utf-8 encoding error: {}", err);
+      HttpResponse::InternalServerError().finish()
+    }
```
Comment (Nutomic) [inline src/prometheus_metrics.rs:90]: Use `?` operator.

File: src/prometheus_metrics.rs [hunk 7]
```diff
@@ -60,63 +62,77 @@ pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyC
...
-fn create_db_pool_metrics() -> DbPoolMetrics {
-  let metrics = DbPoolMetrics {
-    max_size: Gauge::with_opts(Opts::new(
-      "lemmy_db_pool_max_connections",
-      "Maximum number of connections in the pool",
-    ))
-    .unwrap(),
-    size: Gauge::with_opts(Opts::new(
+impl DbPoolMetrics {
+  fn new(context: Arc<PromContext>) -> Self {
```
Comment (dessalines) [inline src/prometheus_metrics.rs:89]: Pry best for this to return a Result.

File: src/lib.rs [hunk 8]
```diff
@@ -144,8 +140,10 @@ pub async fn start_lemmy_server() -> Result<(), LemmyError> {
     });
   }
 
-  #[cfg(feature = "prometheus-metrics")]
-  serve_prometheus(settings.prometheus.as_ref(), context.clone());
+  match settings.prometheus.is_some() {
+    true => serve_prometheus_metrics(settings.prometheus.as_ref(), context.clone()),
+    false => info!("No Prometheus config provided, will not start Prometheus server"),
+  }
```
Comment (Nutomic) [inline src/lib.rs:146]: Use `if let Some(config) = settings.prometheus`, then you dont need any panic. Also I wouldnt log anything if prometheus config is unset.

File: src/lib.rs (hunk 2 above)
Comment (Nutomic) [reply src/lib.rs:168]: Specifically, return PrometheusMetrics from `serve_prometheus_metrics`. Then you will have an `Option<PrometheusMetrics>`. Use `if let Some(metrics) = metrics { app.wrap(metrics) }`

**Comment by andybug** [comment]
Sorry, I've been dealing with some personal issues and haven't had time to look at this and will be out for a bit more. I'll close this PR and reopen in the future when I have time to clean this up.

Thanks for your feedback, I will get this incorporated someday.

**Comment by thegabriele97** [comment]
Any chance to see this handled again?

**Pull request closed without merging**